gremlin-groovy/src/main/java/org/apache/tinkerpop/gremlin/groovy/jsr223/ParameterizedGroovyTranslator.java

```diff
@@ -0,0 +1,340 @@
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package org.apache.tinkerpop.gremlin.groovy.jsr223;
+
+import org.apache.commons.configuration.ConfigurationConverter;
+import org.apache.tinkerpop.gremlin.process.traversal.*;
+import org.apache.tinkerpop.gremlin.process.traversal.Translator.ScriptTranslator;
+import org.apache.tinkerpop.gremlin.process.traversal.step.TraversalOptionParent;
+import org.apache.tinkerpop.gremlin.process.traversal.strategy.TraversalStrategyProxy;
+import org.apache.tinkerpop.gremlin.process.traversal.util.ConnectiveP;
+import org.apache.tinkerpop.gremlin.process.traversal.util.OrP;
+import org.apache.tinkerpop.gremlin.structure.Edge;
+import org.apache.tinkerpop.gremlin.structure.Element;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+import org.apache.tinkerpop.gremlin.structure.VertexProperty;
+import org.apache.tinkerpop.gremlin.structure.util.StringFactory;
+import org.apache.tinkerpop.gremlin.util.function.Lambda;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Converts bytecode to a Parameterized Groovy string of Gremlin.
+ *
+ * @author Marko A. Rodriguez (http://markorodriguez.com)
+ * @author Stephen Mallette (http://stephen.genoprime.com)
+ * @author Stark Arya (sandszhou.zj@alibaba-inc.com)
+ */
+
+public class ParameterizedGroovyTranslator implements ScriptTranslator {
+    private final String traversalSource;
+    private final TypeTranslator typeTranslator;
+
+    private ParameterizedGroovyTranslator(final String traversalSource, TypeTranslator typeTranslator) {
+        this.traversalSource = traversalSource;
+        this.typeTranslator = typeTranslator;
+    }
+
+    public static final ParameterizedGroovyTranslator of(final String traversalSource) {
+        return of(traversalSource,null);
+    }
+
+    public static final ParameterizedGroovyTranslator of(final String traversalSource, final TypeTranslator typeTranslator) {
+        return new ParameterizedGroovyTranslator(traversalSource,
+                Optional.ofNullable(typeTranslator).orElseGet(DefaultParameterizedTypeTranslator::new));
+    }
+
+    //////
+    public ParameterizedGroovyResult parameterizedTranslate(final Bytecode bytecode) {
```

Comment:
I think that you need to implement this slightly differently. I don't think you want a new `parameterizedTranslate(Bytecode)` method. I think you just want:

```java
public ParameterizedGroovyResult translate(Bytecode)
```

You should just consider a new `Translator` implementation like `ScriptTranslator`:

https://github.com/apache/tinkerpop/blob/199c9106a76b09652b8f4d9cb76cbed7ac7fd418/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/Translator.java#L63

that is:

```java
public interface ScriptTranslator extends Translator<String, ParameterizedGroovyResult> 
```

Personally, I think I'd prefer that `ParameterizedGroovyResult` to be named `ParameterizedScript` (and moved `gremlin-core`) so that it could be more generally used in the future if we liked (`gremlin-python` translation for example) thus:

```java
public interface ScriptTranslator extends Translator<String, ParameterizedScript> 
```